Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,560 @@
+  `Substring` and reduce allocations. However, array in capture types would make
+  a much cleaner type signature.
+- For repetitions or an exact count, e.g. `[a-z]{2}`, it would slightly improve
+  type safety to make its capture type be a homogeneous tuple instead of an
+  array. However, because large homogeneous tuples are difficult to work with
+  and because fixed-count repetitions are syntactically very similar to
+  dynamic-count repetitions which produce an array, an tuple capture may come as
+  a surprise. An array on the other is a better choise for both ease of
+  use and consistency.
+
+#### Alternation
+
+Alternations are used to match one of multiple possible patterns.
+
+```txt
+pattern ::= pattern '|' pattern
+```
+
+If there are one or more capturing groups within an alternation, the resulting
+capture type is an `Alternation` that's generic over each option's underlying
+pattern.
+
+```swift
+/([01]+)|[0-9]+|([0-9A-F]+)/
+// => `Regex<Alternation<(Substring, Void, Substring)>>`
+```
+
+If there are no capturing groups within an alternation the resulting capture type is `Void`.
+
+```swift
+/[01]+|[0-9]+|[0-9A-F]+/
+// => `Regex<Void>`
+```
+
+Nested captures follow the algebra previously described.
+
+```swift
+/([01]+|[0-9]+|[0-9A-F]+)/
+// => `Regex<Substring>`
+/(([01]+)|([0-9]+)|([0-9A-F]+))/
+// => Regex<Substring, Alternation<(Substring, Substring, Substring)>>
+```
+
+At the use site, you often want to be able use `Alternation` like an `enum`.
+Ideally you'd be able to exhaustively switch over all the captures.
+
+Like:
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0 {
+        case let .0(binary):
+            return Int(binary, radix: 2)
+        case let .1(decimal):
+            return Int(decimal, radix: 10)
+        case let .2(hex):
+            return Int(decimal, radix: 16)
+        }
+    }
+```
+
+Or:
+
+```swift
+let number = line
+    .firstMatch(/(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0 {
+        case let .binary(str):
+            return Int(str, radix: 2)
+        case let .decimal(str):
+            return Int(str, radix: 10)
+        case let .hex(str):
+            return Int(str, radix: 16)
+        }
+    }
+```
+
+In the fullness of time, we'd like to design a language feature to support this.
+In the meantime, we'd like to do the best we can and leave the door open for a
+source compatible migration.
+
+With variadic generics we think we can defining the following `Alternation` type.
+
+```swift
+struct Alternation<Captures> { ... }
+
+extension<T...> Alternation where Captures == (T...) {
+    var captures: (T?...) { ... }
+}
+```
+
+Which would support:
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0.captures {
```

Comment:
I think we should go with `optionalProjection` as we discussed earlier, or some better name instead of `captures` for this property. Because `Alternation` is already captures. It's not immediately clear what we are doing by calling `captures` on something that already represents captures.